selectel/resource_selectel_mks_cluster_v1.go

```diff
@@ -0,0 +1,560 @@
+			nodegroupIDs := make(map[string]struct{})
+			for _, v := range oldNGSet.List() {
+				m := v.(map[string]interface{})
+				if id, ok := m["id"]; ok {
+					nodegroupIDs[id.(string)] = struct{}{}
+				}
+			}
+
+			// Create each new nodegroup separately and save its ID in the new nodegroup schema.
+			for _, newNG := range newNGSet.List() {
+				for _, addNG := range nodegroupsToAdd.List() {
+					if addNG.(map[string]interface{})["name"] == newNG.(map[string]interface{})["name"] {
+						opts := expandNodegroupCreateOpts(addNG)
+
+						log.Print(msgCreate(objectClusterNodegroups, opts))
+						_, err := nodegroup.Create(ctx, mksClient, d.Id(), opts)
+						if err != nil {
+							return errCreatingObject(objectClusterNodegroups, err)
+						}
+
+						log.Printf("[DEBUG] waiting for cluster %s to become 'ACTIVE'", d.Id())
+						timeout := d.Timeout(schema.TimeoutUpdate)
+						err = waitForClusterV1ActiveState(ctx, mksClient, d.Id(), timeout)
+						if err != nil {
+							return errCreatingObject(objectClusterNodegroups, err)
+						}
+
+						log.Print(msgGet(objectClusterNodegroups, d.Id()))
+						allNodegroups, _, err := nodegroup.List(ctx, mksClient, d.Id())
+						if err != nil {
+							return errGettingObject(objectClusterNodegroups, d.Id(), err)
+						}
+
+						for _, n := range allNodegroups {
+							if _, ok := nodegroupIDs[n.ID]; !ok {
+								newNG.(map[string]interface{})["id"] = n.ID
+								nodegroupIDs[n.ID] = struct{}{}
+							}
+						}
+					}
+				}
+			}
+			d.Set("nodegroups", newNGSet)
+		}
+
+		for _, ng := range nodegroupsToRemove.List() {
+			if nodegroupID, ok := ng.(map[string]interface{})["id"]; ok {
+				log.Print(msgDelete(objectClusterNodegroups, d.Id()))
+				_, err := nodegroup.Delete(ctx, mksClient, d.Id(), nodegroupID.(string))
+				if err != nil {
+					return errDeletingObject(objectClusterNodegroups, d.Id(), err)
+				}
+
+				timeout := d.Timeout(schema.TimeoutDelete)
+				err = waitForClusterV1ActiveState(ctx, mksClient, d.Id(), timeout)
+				if err != nil {
+					return errDeletingObject(objectClusterNodegroups, d.Id(), err)
+				}
+			}
+		}
+	}
+
+	return resourceMKSClusterV1Read(d, meta)
+}
+
+func resourceMKSClusterV1Delete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	resellV2Client := config.resellV2Client()
+	ctx := context.Background()
+	tokenOpts := tokens.TokenOpts{
+		ProjectID: d.Get("project_id").(string),
+	}
+
+	log.Print(msgCreate(objectToken, tokenOpts))
+	token, _, err := tokens.Create(ctx, resellV2Client, tokenOpts)
+	if err != nil {
+		return errCreatingObject(objectToken, err)
+	}
+
+	region := d.Get("region").(string)
+	endpoint := getClusterV1Endpoint(region)
+	mksClient := v1.NewMKSClientV1(token.ID, endpoint)
+
+	log.Print(msgDelete(objectCluster, d.Id()))
+	_, err = cluster.Delete(ctx, mksClient, d.Id())
+	if err != nil {
+		return errDeletingObject(objectCluster, d.Id(), err)
+	}
+
```

Comment:
We should wait for cluster to become deleted here by waiting for 404 status code from MKS API.